My Simple App/Business Logic/AddressStore/AddressStoreImpl.swift

```diff
@@ -0,0 +1,123 @@
+//
+//  AddressStoreImpl.swift
+//  My Simple App
+//
+//  Created by Erick Sanchez on 8/12/20.
+//  Copyright © 2020 Erick Sanchez. All rights reserved.
+//
+
+import Foundation
+
+func injectAddressStore() -> AddressStore {
+  AddressStoreImpl(internalUserDefaults: injectInternalUserDefaults())
+}
+
+class AddressStoreImpl: AddressStore {
+
+  // TODO: mock dependencies and write unit tests.
+
+  private static let userAddressesKey = "USER_ADDRESSES"
+
+  fileprivate let storeDidChange = Event<Void>()
+
+  fileprivate let internalUserDefaults: InternalUserDefaults
+
+  init(internalUserDefaults: InternalUserDefaults) {
+    self.internalUserDefaults = internalUserDefaults
+  }
+
+  func addresses() -> AllAddressesResource {
+    return AnyReactableResource(base: AllAddressesReactableResource(store: self))
+  }
+
+  func storeAddress(name: String) {
+    let addressObject = AddressObject(
+      identifier: UUID(),
+      name: name,
+      longitude: nil,
+      latitude: nil,
+      temperature: nil,
+      iconName: nil
+    )
+    let newAddresses = [addressObject] + self.loadData()
+    self.saveData(addresses: newAddresses)
+  }
+
+  func storeAddress(name: String, latitude: Double, longitude: Double, weather: Weather) {
+    let addressObject = AddressObject(
+      identifier: UUID(),
+      name: name,
+      longitude: longitude,
+      latitude: latitude,
+      temperature: weather.temperature,
+      iconName: weather.iconName
+    )
+    let newAddresses = [addressObject] + self.loadData()
+    self.saveData(addresses: newAddresses)
+  }
+
+  func delete(address: Address) {
+    let idMatcher: (AddressObject) -> Bool = { $0.identifier == address.identifier }
+    var addressesData = self.loadData()
+
+    guard let indexToRemove = addressesData.firstIndex(where: idMatcher) else {
+      return assertionFailure("Failed to delete given address, not found in database: \(address)")
+    }
+
+    addressesData.remove(at: indexToRemove)
+    self.saveData(addresses: addressesData)
+  }
+
+  // MARK: - Private
+
+  // TODO: load and save off the main thread.
+
+  private func saveData(addresses: [AddressObject]) {
+    do {
+      let newAddressesData = try JSONEncoder().encode(addresses)
+      self.internalUserDefaults.set(newAddressesData, forKeyPath: Self.userAddressesKey)
```

Comment:
woodward4422 2 days ago @woodward4422
Why use user defaults with these data models? I feel like this is not what User Defaults was designed for, IMO User Defaults should only be used for small app configurations and not for saving data models. Why not use either Core Data or realm for this?

LinnierGames 40 minutes ago Author Owner
Using UserDefaults was in favor of using common iOS tools. I agree UserDefaults should be used with more lightweight stuff like preferences but this beats using something like setting up CoreData or FileManager and file paths

Realm was not used bc it's a 3rd party tool and (for now) I've mentioned to only use first-party tools